### PR TITLE
trace: trace to stderr

### PR DESCRIPTION
**Pull request by oxij**
It's a bad idea to write debug output to stdout, it gets lost there.

### CODE REVIEW

File: src/trace.c [hunk 1]
```diff
@@ -78,12 +78,12 @@ trace_error(const char *fmt, ...)
     vfprintf(stderr, fmt, args);
     va_end(args);
 
-    if (!stdout_is_a_tty) {
-        // If stdout is redirected to a file, make sure it also gets error messages.
+    if (!stderr_is_a_tty) {
+        // If stderr is redirected to a file, make sure it also gets error messages.
         // That helps to figure out where errors were.
-        fprintf(stdout, "[apulse] [error] ");
+        fprintf(stderr, "[apulse] [error] ");
```
Comment (i-rinat) [inline src/trace.c]: This duplicates error messages. There is already `fprintf()` that writes to stderr half a dozen lines earlier.

**Comment by oxij** [comment]
> This duplicates error messages. There is already `fprintf()` that writes to stderr half a dozen lines earlier.

Oops, right.

> being sent to the stderr, as well as to the stdout.

Yes, this is incorrect in that it duplicates the messages. I made this patch with sed :)

However, my problem with the state before this patch is that since some messages get sent stdout instead of stderr apps that draw with ncurses get pretty ugly with tracing enabled and redraws erase the debugging info them from the view.

I think everything should just be sent to stderr.

**Comment by oxij** [comment]
Also, a little thing I made you might be interested in: https://github.com/oxij/libcardiacarrest
Unrelated to this PR but strongly related to `apulse`.

**Comment by i-rinat** [comment]
I don't like the idea of using stderr exclusively. It definitely will be an inconvenience for me during coding/debugging.

But still, I pushed 0bd2288d7cae26876ecb44e22d9f188dee6c41d5 to the `master` branch. It's not what you wanted to merge, but if you want to log to the `stderr`, you need to patch only the line where `log_to_stderr` flag is set.

I think it's possible to make it configurable at run time. Haven't decided how yet.

**Comment by i-rinat** [comment]
> you might be interested in: https://github.com/oxij/libcardiacarrest

Sometimes I regret I made `apulse` available. For some reason some people believe that I share their hatred towards PulseAudio. Which is not the case.

**Comment by oxij** [comment]
> I think it's possible to make it configurable at run time. Haven't decided how yet.

I would be perfectly happy with an official compile time option I can use when packaging, like "-DLOG_TO_STDERR=1" or something similar.

>> you might be interested in: https://github.com/oxij/libcardiacarrest
>
> Sometimes I regret I made `apulse` available. For some reason some people believe that I share their hatred towards PulseAudio. Which is not the case.

I find it hard to believe that one can read PA's source and still be like "nah, its fine, it totally makes sense". But, well, okay, opinion noted, I guess.

I think `libcardiacarrest` provides useful service even if you don't hate PA, though.

**Comment by i-rinat** [comment]
> like "-DLOG_TO_STDERR=1"

Why not. Added in 0be73093ee7996d4470c01fd152029bdf61ddaf0.

> I find it hard to believe that one can read PA's source and still be like "nah, its fine, it totally makes sense".

You didn't look into apulse's source code, did you? :grinning:

**Comment by oxij** [comment]
Rinat Ibragimov <notifications@github.com> writes:

> Why not. Added in 0be73093ee7996d4470c01fd152029bdf61ddaf0.

Perfect. Thank you. You can close this, or I will the next time I login into GitHub.

> You didn't look into apulse's source code, did you? :grinning:

I did. The lack of library-to-daemon protocol handling makes this considerably better.

**Pull request closed without merging**